### DISCUSSION TITLE
Creating a Filament Drying button

### DISCUSSION THREAD

**Discussion by Manu-pardo** [KlipperScreen/KlipperScreen#1489] [Q&A]
Hi. I've been using klipperscreen for a while, but now I want to do some modifications that require more coding experience (I have very limited knowledge in python). One of those modifications is adding a menu button called "Filament Drying", similar to the option Bambu Lab printers have.  I have the following custom macros that I use in Mainsail:

```
;An idea for using the heated bed of a 3D printer as a filament dryer.
; Adds GCODE command: START_DRYER TIME=T TEMPERATURE=C
; (T is time in seconds, C is bed temperature in Celsuis)
; To stop drying early, use STOP_DRYER.
; Also defined some utility macros: DRY_PLA, DRY_PETG and DRY_ABS.
; Edit these with your own preferred defaults.

[gcode_macro START_DRYER]
description: Start the heated bed filament dryer.
gcode:
    #{% set ChamberTemperature = params.CHAMBER | default(25.0) | float %}
    {% set BedTemperature = params.TEMPERATURE | default(50.0) | float %}
    {% set DryTime = params.TIME | default(14400) | int %}
    ; turn the heaters on, however you do that.
    M140 S{BedTemperature}      ; Sets the print bed temperature without waiting.
    #M141 S{ChamberTemperature}  ; [OPTIONAL] Sets the enclosure temperature.
    ; then finally,
    SET_GCODE_VARIABLE MACRO=DRYER_STATUS VARIABLE=time_remaining VALUE={DryTime}
    SET_GCODE_VARIABLE MACRO=DRYER_STATUS VARIABLE=bed_temperature VALUE={BedTemperature}
    #SET_GCODE_VARIABLE MACRO=DRYER_STATUS VARIABLE=chamber_temperature VALUE={ChamberTemperature}
    UPDATE_DELAYED_GCODE ID=DRYER_TIMER DURATION=1

[gcode_macro STOP_DRYER]
gcode:
    ; Turn off heaters etc. here
    M140 S0 ; Disable bed heater
    #M141 S0 ; [OPTIONAL] Disable enclosure heater/fan
    SET_GCODE_VARIABLE MACRO=DRYER_STATUS VARIABLE=time_remaining VALUE=0
    UPDATE_DELAYED_GCODE ID=DRYER_TIMER DURATION=0    ; Stop the timer.
    M117 Drying Stopped

[gcode_macro DRYER_STATUS]
variable_time_remaining: 0
variable_bed_temperature: 0
#variable_chamber_temperature: 0
gcode:
    {% if time_remaining > 0 %}
        M140 S{bed_temperature} ; Reset bed temperature (prevents timeout)
        SET_GCODE_VARIABLE MACRO=DRYER_STATUS VARIABLE=time_remaining VALUE={time_remaining - 1}
        M117 Drying {time_remaining}
    {% else %}
        STOP_DRYER
    {% endif %}

[delayed_gcode DRYER_TIMER]
gcode:
    UPDATE_DELAYED_GCODE ID=DRYER_TIMER DURATION=1
    DRYER_STATUS

[gcode_macro DRY_PLA]
gcode:
    START_DRYER TEMPERATURE=50 TIME=14400

[gcode_macro DRY_PETG]
gcode:
    START_DRYER TEMPERATURE=55 TIME=18000

[gcode_macro DRY_ABS]
gcode:
    START_DRYER TEMPERATURE=65 TIME=14400

```

I temporarily made a button that has the options for ABS, PLA and PETG,  but I am unable to create a button that could allow me to set both time and temperature for ```[gcode_macro START_DRYER]``` due to my lack of expertise. Therefore, I am looking for help to see if it is possible. I know that it should be better to make a dedicated gcode macro only for Klipperscreen (and to leave the shown macros only for Mainsail's GUI).

By the way, I am temporarily using three buttons inside my custom menu with the following added to klipperscreen.conf:

```
[menu __main filament_drying]
name: Filament Drying
icon: filament_drying

[menu __main filament_drying PLA]
name: Dry PLA
method: printer.gcode.script
params: {"script":"DRY_PLA"}

[menu __main filament_drying ABS]
name: Dry ABS
method: printer.gcode.script
params: {"script":"DRY_ABS"}

[menu __main filament_drying PETG]
name: Dry PETG
method: printer.gcode.script
params: {"script":"DRY_PETG"}

```

Pd: I designed a custom icon for my menu:

![filament_drying](https://github.com/user-attachments/assets/db860a3e-f36e-4fbc-a07a-2444dcd5b788)

Any help would be much appreciated. If this works, I might create a simple repo with instructions to implement this menu button, as many of us don't have dedicated filament dryers .

**Comment by alfrix** [comment]
> I am unable to create a button that could allow me to set both time and temperature for [gcode_macro START_DRYER]

something like this should work
```
[menu __main filament_drying PETG]
name: Dry PETG
method: printer.gcode.script
params: {"script":"START_DRYER TEMPERATURE=55 TIME=18000"}
```

and use the macros panel to execute START_DRYER when you want to set speficif values.

probably this would be handy too
```
[menu __main filament_drying STOP]
name: Stop
method: printer.gcode.script
params: {"script":"STOP_DRYER"}
```
i would leave it like that, since that might be good enough

there are other posibilities like modifying the macros to use time and temp as variables that can be set from other macros (using SET_GCODE_VARIABLE), and then create a set_time and set_temp menus with a [prompt](https://klipperscreen.github.io/KlipperScreen/macros/#prompts) or submenu with discrete options, and the start as a separate button that reads the variables.

or creating a full custom panel, but for that you need coding with python and pygobject.
